### PR TITLE
Fix lazy enumerable typings Fixes #2048

### LINKED ISSUE #2048: Lazy Enumerable Typed Incorrectly
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0AT.reveal_type(%5B1%5D.lazy.select(%26%3Aeven%3F)))

#### Observed output
```ruby
# typed: true
T.reveal_type([1].lazy.select(&:even?))
# => Revealed type: T::Array[Integer] 
irb(main):001:0> [1].lazy.select(&:even?).class
# => Enumerator::Lazy
```

#### Expected behavior

The [Enumerable RBI](https://github.com/sorbet/sorbet/blob/fd537c1af87d87915198ed1808a82ac30f66109a/rbi/core/enumerable.rbi) file contains methods definitions which all return arrays of `Elem`. Most of the methods, actually return another Lazy Enumerable. (#chunk, #chunk_while, #collect, #collect_concat, #drop, #drop_while, #enum_for, #find_all, #flat_map, #grep, #grep_v, #lazy, #map, #reject, #select, #slice_after, #slice_before, #slice_when, #take, #take_while, #to_enum, #uniq, #zip) 

I should be able to make this change if it only requires making a duplicate `Enumerable` module returning the correct values and mixing it in [here](https://github.com/sorbet/sorbet/blob/c843cfad7156eaea0abc27e21e10154266b3d557/rbi/core/enumerator.rbi#L413). I'd just like to ensure I make an acceptable change before I submit a PR.

### PR DESCRIPTION
**Pull request by bmalinconico**
Fixes #2048

This change adds new method signatures to the [Enumerable::Lazy](https://ruby-doc.org/core-2.2.10/Enumerator/Lazy.html) class. Every one of these methods is identical to it's Enumerable method, but it returns a new lazy enumerable.

The following methods were not added since I didn't see them in Enumerable itself. If they are desirable I can add them in either this PR or a second PR.

* slice_after(*args)
* chunk
* slice_before
* slice_when
* zip
* to_enum

One sig I changed from the Enumerable module was for `take`. The sig defined in Enumerable is

```ruby
  sig do
    params(
        n: Integer,
    )
    .returns(T.nilable(T::Array[Elem]))
  end
  def take(n); end
```
The [docs](https://ruby-doc.org/core-2.6.3/Enumerable.html#method-i-take) do not appear to indicate that nil is a valid return value, and I am unable to get a nil return value in a console.

Negative values throw errors and  0 returns an empty array.

Since I belive this typing to be in error I corrected it, but if you are aware of a way to get `take` to return `nil`, I would like to verify the Lazy `take` typing to verify its behavior.

### Motivation
The typings are wrong, and I love lazy enumerables and I want to love Sorbet.

### Test plan
I'm a bit unsure of the overall Sorbet testing philosophy for RBIs. I included an `assert_type!` for every method signature added.

See included automated tests.

### CODE REVIEW

**Comment by elliottt** [comment]
It looks like there are a few places where you used `Enumerable::Lazy` instead of `Enumerator::Lazy`.

File: rbi/core/enumerator.rbi [hunk 1]
```diff
@@ -413,6 +413,247 @@ end
...
+  #
+  # ```ruby
+  # a = [1, 2, 3, 4, 5, 0]
+  # a.drop(3)             #=> [4, 5, 0]
+  # ```
+  sig do
+    params(
+        n: Integer,
+    )
+    .returns(Enumerable::Lazy[Elem])
```
Comment (elliottt) [inline rbi/core/enumerator.rbi]:
```suggestion
    .returns(Enumerator::Lazy[Elem])
```

**Comment by bmalinconico** [comment]
I'll review the failing specs tomorrow, thanks for the feedback

**Comment by bmalinconico** [comment]
I've reviewed the remaining failures.

flat_map is returning the incorrect signature, I assume due to the note left in the `rbi/core/enumerable.rbi`. Currently it is returning `T::Array[Enumerator::Lazy[[String]]]`, the correct typing should be `Enumerator::Lazy[String]`.

The correct typing should be the block returns an Enumerable of type U, the return value of flat_map should be `Enumerable::Lazy[U]`

What is the preferred way to approach the current limitations. I also tried pulling the type out of T::Array, but that seems to fail too.

**Comment by elliottt** [comment]
I've unblocked CI so that I can get a look at the remaining errors. Which is the comment you're referring to in `rbi/core/enumerable.rbi`?

**Comment by bmalinconico** [comment]
@elliottt 

```
  ### N.B. this signature is wrong; Our generic method implementation
  ### cannot model the correct signature, so we pass through the return
  ### type of the block and then fix it up in an ad-hoc way in Ruby. A
  ### more-correct signature might be:
  ###   sig do
  ###     type_parameters(:U).params(
  ###       blk: T.proc.params(arg0: Elem).returns(T.any(T::Array[T.type_parameter(:U)], T.type_parameter(:U)),
  ###     )
  ###     .returns(T.type_parameter(:U))
  ###   end
  ###
  ### But that would require a lot more sophistication from our generic
  ### method inference.

  # Returns a new array with the concatenated results of running *block* once
  # for every element in *enum*.
  #
  # If no block is given, an enumerator is returned instead.
  #
  # ```ruby
  # [1, 2, 3, 4].flat_map { |e| [e, -e] } #=> [1, -1, 2, -2, 3, -3, 4, -4]
  # [[1, 2], [3, 4]].flat_map { |e| e + [100] } #=> [1, 2, 100, 3, 4, 100]
  # ```
  sig do
    type_parameters(:U).params(
        blk: T.proc.params(arg0: Elem).returns(T.type_parameter(:U)),
    )
    .returns(T.type_parameter(:U))
  end
  sig {returns(T::Enumerator[Elem])}
  def flat_map(&blk); end
```

of the remaining failures, i'm unsure of the source for this one 

`https://github.com/sorbet/sorbet/tree/master/rbi/core/enumerator.rbi#L660: Did you mean: Enumerator::Yielder`

I'll also review the typings for `collect_concat` on my next pass

**Comment by elliottt** [comment]
The tests can be easier to debug if you use `T.reveal_type` and put the expected failure in a comment that trails that statement. Here's an example:
https://github.com/sorbet/sorbet/blob/master/test/testdata/call_with_block.rb#L46

The error about the Yielder recommendation is new, just because the line in the rbi file changed. If you run `tools/scripts/update_exp_files.sh` it will update the expectation file.

**Comment by bmalinconico** [comment]
@elliottt There is one set of pending tests, and they are failing due to some interpreter code intercepting enumerable flat map. Here is a [playground link](https://sorbet.run/#class%20Enumerator%3A%3ALazy%0A%20%20extend%20T%3A%3ASig%0A%20%20sig%20do%0A%20%20%20%20type_parameters(%3AU).params(%0A%20%20%20%20%20%20%20%20blk%3A%20T.proc.params(arg0%3A%20Elem)%0A%20%20%20%20%20%20%20%20.returns(T%3A%3AEnumerable%5BT.type_parameter(%3AU)%5D)%2C%0A%20%20%20%20)%0A%20%20%20%20.returns(Enumerator%3A%3ALazy%5BT.type_parameter(%3AU)%5D)%0A%20%20end%0A%20%20sig%20%7Breturns(Enumerator%3A%3ALazy%5BElem%5D)%7D%0A%20%20def%20flat_map(%26blk)%3B%20end%0Aend%0A%0AT.reveal_type%20%5B1%5D.lazy.flat_map%7B%7Ci%7C%20%5Bi.to_s%5D%20%7D) demonstrating the current output.

This is happening due to [this code](https://github.com/sorbet/sorbet/blob/master/infer/environment.cc#L740). Unfortunately I do not believe I will be able to correct this issue on my own, but it looks like it needs to be updated to account for lazy enumerations.

Let me know your preferred approach for this, I think removing or commenting out the flat map declaration would be preferable to leaving it, since the correct typing is _closer_ to T::Array[T.type_parameter] than T::Array[Enumerable::Lazy[T.type_parameter]].

**Comment by elliottt** [comment]
Thanks for the update,  I'll investigate the issues with flat_map type handling 👍

**Comment by elliottt** [comment]
I pushed a fix to your branch that adds another exception for `Enumerable::Lazy` receivers in `flatmapHack`, and fixed the tests for `flat_map`.

File: core/SymbolRef.h [hunk 2]
```diff
@@ -397,6 +397,10 @@ class Symbols {
         return SymbolRef(nullptr, 83);
     }
 
+    static SymbolRef Lazy() {
```
Comment (ptarjan) [inline core/SymbolRef.h]: can you name this `Enumerator_Lazy` please?  The convention is to name it as close to the full path as makes sense

**Comment by ptarjan** [comment]
@elliottt you can either merge it as is and fix if yourself after, ask the author to rename it, or push a fix to this branch

File: core/SymbolRef.h (hunk 2 above)
Comment (elliottt) [reply core/SymbolRef.h]: Done 👍

**Comment by bmalinconico** [comment]
I wanted to verify that you are no longer waiting on my for this PR

**Comment by elliottt** [comment]
@bmalinconico you're all set, thank you for checking in. This PR raised a couple of errors on the main stripe codebase, so I just need to  dig into those before I merge this.

**Pull request merged**